core/plugins/org.polarsys.capella.core.sirius.analysis/src/org/polarsys/capella/core/sirius/analysis/EventContextServices.java

```diff
@@ -0,0 +1,414 @@
+        //   EventContext(IR3, e3, false, 2)
+        // @formatter:on
+  }
+
+  /**
+   * Compute EventStructure for InteractionState.
+   * 
+   * @param ancestors
+   *          Deque<EObject>
+   * @param result
+   *          List<EventContext>
+   * @param end
+   *          InteractionState
+   */
+  private static void visit(Deque<CapellaElement> ancestors, List<EventContext> result, InteractionState end,
+      Optional<TimeLapse> timeLapse) {
+    if (timeLapse.isPresent() && timeLapse.get() instanceof StateFragment
+        && ((StateFragment) timeLapse.get()).getStart() == end) {
+      result.add(new EventContext(ancestors.peek(), timeLapse.get(), true, ancestors.size() + 1));
+    }
+
+    if (timeLapse.isPresent() && timeLapse.get() instanceof StateFragment
+        && ((StateFragment) timeLapse.get()).getFinish() == end) {
+      result.add(new EventContext(ancestors.peek(), timeLapse.get(), false, ancestors.size() + 1));
+    }
+  }
+
+  /**
+   * Compute EventStructure for ExecutionEnd.
+   * 
+   * @param ancestors
+   *          Deque<EObject>
+   * @param result
+   *          List<EventContext>
+   * @param end
+   *          ExecutionEnd
+   */
+  private static void visit(Deque<CapellaElement> ancestors, List<EventContext> result, ExecutionEnd end,
+      Optional<TimeLapse> timeLapse) {
+    // Should not happen in current implementation of Capella Sequence diagrams
+    if (timeLapse.isPresent() && timeLapse.get().getStart() == end) {
+      result.add(new EventContext(ancestors.peek(), timeLapse.get(), true, ancestors.size() + 1));
+      ancestors.push(timeLapse.get());
+    }
+
+    // Execution without return branch end / ASyncCall
+    if (timeLapse.isPresent() && timeLapse.get().getFinish() == end) {
+      ancestors.pop();
+      result.add(new EventContext(ancestors.peek(), timeLapse.get(), false, ancestors.size() + 1));
+    }
+  }
+
+  /**
+   * Return covered instance role.
+   * 
+   * @param end
+   *          InteractionFragment
+   * @return covered instance role.
+   */
+  private static InstanceRole getCoveredInstanceRole(InteractionFragment end) {
+    InstanceRole covered = null;
+    if (end instanceof AbstractEnd) {
+      covered = getInteractionCache(AbstractEnd::getCovered, (AbstractEnd) end);
+    } else if (end instanceof InteractionState) {
+      covered = getInteractionCache(InteractionState::getCovered, (InteractionState) end);
+    }
+    return covered;
+  }
+
+  /**
+   * Compute cache for TimeLapse start and finish ends.
+   * 
+   * @param scenario
+   *          Scenario
+   */
+  private static void computeTimeLapseStartAndEndCache(Scenario scenario) {
+    // cache Execution.start -> Execution and Execution.finish -> Execution
+    // scan timelapse only one time
+    if (getInteractionFragmentToTimeLapseCache().isEmpty()) {
+      Stream<TimeLapse> timeLapses = scenario.getOwnedTimeLapses().stream()
+          .filter(tl -> !(tl instanceof AbstractFragment));
```

Comment:
Maybe AbstractFragment could be taken into account in a future enhancement since it's absolutely suitable to use the same logic of EventContext to process elements such as CombinedFragment and InteractionUse.

Reply:
Ok i see